src/udev/udev-builtin-path_id.c

```diff
@@ -203,10 +201,25 @@ static sd_device *handle_scsi_sas(sd_device *parent, char **path) {
         }
 
         format_lun_number(parent, &lun);
-        if (sas_address)
+        if (sas_address) {
+                /* Use expander-side phy id */
+                if (sd_device_get_sysname(port, &sysname) < 0)
+                        return NULL;
+                if (!strneq(sysname, "port-", STRLEN("port-")))
+                        return NULL;
+                if (!snprintf_ok(phyname, sizeof phyname, "phy-%s", sysname + STRLEN("port-")))
+                        return NULL;
+                if (sd_device_new_from_subsystem_sysname(&port_sasdev, "sas_phy", phyname) < 0)
+                        return NULL;
+                if (sd_device_get_sysattr_value(port_sasdev, "phy_identifier", &phy_id) < 0)
+                        return NULL;
                  path_prepend(path, "sas-exp%s-phy%s-%s", sas_address, phy_id, lun);
-        else
-                 path_prepend(path, "sas-phy%s-%s", phy_id, lun);
+        } else {
+                /* Get connected phy */
+                if (sd_device_get_sysattr_value(target_sasdev, "phy_identifier", &phy_id) < 0)
+                        return NULL;
+                path_prepend(path, "sas-phy%s-%s", phy_id, lun);
```

Comment:
I am not sure why this case is different. HBA port may have multiple phys and each phy may connect to different SAS disk so it should have exactly the same problem.

Reply:
Good point, and it makes sense to fix up both cases at once.